bakery/codec.go

```diff
@@ -85,143 +91,218 @@ func encodeCaveatIdV1(
 	return buf, nil
 }
 
+// encodeCaveatV2 creates a version 2 third-party caveat.
+func encodeCaveatV2(
+	condition string,
+	rootKey []byte,
+	thirdPartyPubKey *PublicKey,
+	key *KeyPair,
+) ([]byte, error) {
+	return encodeCaveatV2V3(Version2, condition, rootKey, thirdPartyPubKey, key, nil)
+}
+
+// encodeCaveatV3 creates a version 3 third-party caveat.
+func encodeCaveatV3(
+	condition string,
+	rootKey []byte,
+	thirdPartyPubKey *PublicKey,
+	key *KeyPair,
+	ns *checkers.Namespace,
+) ([]byte, error) {
+	return encodeCaveatV2V3(Version3, condition, rootKey, thirdPartyPubKey, key, ns)
+}
+
 const publicKeyPrefixLen = 4
 
-// encodeCaveatIdV2 creates a version 2 third-party caveat.
+// version3CaveatMinLen holds an underestimate of the
+// minimum length of a version 3 caveat.
+const version3CaveatMinLen = 1 + 4 + 32 + 24 + box.Overhead + 1
+
+// encodeCaveatV3 creates a version 2 or version 3 third-party caveat.
 //
-// The v2 format has the following packed binary fields:
-// version 2 [1 byte]
-// first 4 bytes of third-party Curve25519 public key [4 bytes]
-// first-party Curve25519 public key [32 bytes]
-// nonce [24 bytes]
-// encrypted secret part [rest of message]
-func encodeCaveatIdV2(
+// The format has the following packed binary fields (note
+// that all fields up to and including the nonce are the same
+// as the v2 format):
+//
+// 	version 2 or 3 [1 byte]
+// 	first 4 bytes of third-party Curve25519 public key [4 bytes]
+// 	first-party Curve25519 public key [32 bytes]
+// 	nonce [24 bytes]
+// 	encrypted secret part [rest of message]
+//
+// The encrypted part encrypts the following fields
+// with box.Seal:
+//
+// 	version 2 or 3 [1 byte]
+// 	length of root key [n: uvarint]
+// 	root key [n bytes]
+// 	length of encoded namespace [n: uvarint] (Version 3 only)
+// 	encoded namespace [n bytes] (Version 3 only)
+// 	condition [rest of encrypted part]
+func encodeCaveatV2V3(
+	version Version,
 	condition string,
 	rootKey []byte,
 	thirdPartyPubKey *PublicKey,
 	key *KeyPair,
+	ns *checkers.Namespace,
 ) ([]byte, error) {
-	var nonce [NonceLen]byte
-	if _, err := rand.Read(nonce[:]); err != nil {
-		return nil, errgo.Notef(err, "cannot generate random number for nonce")
+
+	var nsData []byte
+	if version >= Version3 {
+		data, err := ns.MarshalText()
+		if err != nil {
+			return nil, errgo.Mask(err)
+		}
+		nsData = data
 	}
-	data := make([]byte, 0, 1+publicKeyPrefixLen+KeyLen+NonceLen+1+binary.MaxVarintLen64+len(rootKey)+len(condition)+box.Overhead)
-	data = append(data, byte(Version2)) //version
+	// dataLen is our estimate of how long the data will be.
+	// As we always use append, this doesn't have to be strictly
+	// accurate but it's nice to avoid allocations.
+	dataLen :=
+		1 + // version
+			publicKeyPrefixLen +
```

Comment:
The formatting here seems a little odd. I'm willing to believe gofmt did it for you though.